staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go

```diff
@@ -101,6 +103,75 @@ func (s *Downloader) Download(handler http.Handler, etag string) (returnSpec *sp
 	}
 }
 
+// gvList is a struct for the response of the /openapi/v3 endpoint to unmarshal into
+type gvList struct {
+	Paths []string `json:"Paths"`
+}
+
+// SpecETag is a OpenAPI v3 spec and etag pair for the endpoint of each OpenAPI group/version
+type SpecETag struct {
+	spec *spec3.OpenAPI
+	etag string
+}
+
+// DownloadV3 downloads OpenAPI v3 for all groups of a given handler
+func (s *Downloader) DownloadV3(handler http.Handler) (returnSpec map[string]*SpecETag, err error) {
+	// TODO(jefftree): https://github.com/kubernetes/kubernetes/pull/105945#issuecomment-966455034
+	// Move to proxy request in the aggregator and let the APIServices serve the OpenAPI directly
+	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
+	handler = http.TimeoutHandler(handler, specDownloadTimeout, "request timed out")
+
+	req, err := http.NewRequest("GET", "/openapi/v3", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Accept", "application/json")
```

Comment:
Should we use the `If-None-Match` that v2 uses? That would avoid the transfer altogether. Also possibly as an optimization later, we should use v3 since it's much much faster (assuming all of that doesn't go away when we proxy the request).

Reply:
Ah yes, that'll save quite a bit of transfers. Will add the header